plugins/modem-manager/fu-firehose-updater.c

```diff
@@ -0,0 +1,699 @@
+					   " SkipStorageInit=\"%d\""
+					   "/>",
+					   CONFIGURE_MEMORY_NAME,
+					   CONFIGURE_VERBOSE,
+					   CONFIGURE_ALWAYS_VALIDATE,
+					   CONFIGURE_MAX_DIGEST_TABLE_SIZE_IN_BYTES,
+					   max_payload_size,
+					   CONFIGURE_ZLP_AWARE_HOST,
+					   CONFIGURE_SKIP_STORAGE_INIT);
+		cmd_bytearray = g_byte_array_new_take ((guint8 *)cmd_str, strlen (cmd_str));
+
+		if (!fu_firehose_updater_send_and_receive (self, cmd_bytearray, &rsp_silo, &rsp_node, error)) {
+			g_prefix_error (error, "Failed to run configure command: ");
+			return 0;
+		}
+
+		/* retry if we're told to use a different max payload size */
+		if (!fu_firehose_updater_check_operation_result (rsp_node, NULL)) {
+			guint64 suggested_max_payload_size;
+			g_autoptr(XbNode) root = NULL;
+
+			root = xb_silo_get_root (rsp_silo);
+			suggested_max_payload_size = xb_node_get_attr_as_uint (root, "MaxPayloadSizeToTargetInBytes");
+			if ((suggested_max_payload_size > G_MAXINT) ||
+			    ((gint)suggested_max_payload_size == max_payload_size)) {
+				break;
+			}
+
+			suggested_max_payload_size = max_payload_size;
+			continue;
+		}
+
+		/* if operation is successful, return the max payload size we requested */
+		return max_payload_size;
+	}
+
+	g_set_error_literal (error,
+			     G_IO_ERROR,
+			     G_IO_ERROR_FAILED,
+			     "Configure operation failed");
+	return 0;
+}
+
+gboolean
+fu_firehose_updater_reset (FuFirehoseUpdater *self, GError **error)
+{
+	const gchar *cmd_str = "<power value=\"reset\" />";
+	GByteArray *cmd_bytearray = NULL;
+	g_autoptr(XbSilo) rsp_silo = NULL;
+	g_autoptr(XbNode) rsp_node = NULL;
+
+	cmd_bytearray = g_byte_array_append (g_byte_array_new (), (const guint8 *)cmd_str, strlen (cmd_str));
+	if (!fu_firehose_updater_send_and_receive (self, cmd_bytearray, &rsp_silo, &rsp_node, error)) {
+		g_prefix_error (error, "Failed to run reset command: ");
+		return FALSE;
+	}
+
+	if (!fu_firehose_updater_check_operation_result (rsp_node, NULL)) {
+		g_set_error_literal (error, G_IO_ERROR, G_IO_ERROR_FAILED,
+				     "Reset operation failed");
+		return FALSE;
+	}
+
+	/* read out all of the remaining messages. otherwise modem won't go into reset */
+	while (fu_firehose_updater_send_and_receive (self, NULL, &rsp_silo, &rsp_node, error));
+
+	return TRUE;
+}
+
+static gboolean
+fu_firehose_updater_send_program_file (FuFirehoseUpdater *self, const gchar *program_filename,
+				       GBytes *program_file, guint payload_size, guint sector_size,
+				       GError **error)
+{
+	gsize program_file_size = g_bytes_get_size (program_file);
+	gssize total_pending_size = (gssize) program_file_size;
+	guint n_blocks;
+	g_autofree guint8 *last_block = NULL;
+
+	/* helper for the last block, which may need padding up to the next
+	 * sector size */
+	g_warn_if_fail (payload_size % sector_size == 0);
+	last_block = g_malloc0 (payload_size);
+
+	n_blocks = (guint) (program_file_size / payload_size);
+	if (program_file_size % payload_size != 0)
+		n_blocks++;
```

Comment:
isn't FuChunk going to be more helpful here?

Reply:
did not know about it, rewrote the function with FuChunk